integrations/crowdnode/src/main/java/org/dash/wallet/integrations/crowdnode/ui/CrowdNodeViewModel.kt

```diff
@@ -17,13 +17,127 @@
-class CrowdNodeViewModel @Inject constructor() : ViewModel() {
+class CrowdNodeViewModel @Inject constructor(
+    private val config: Configuration,
+    private val walletDataProvider: WalletDataProvider,
+    private val crowdNodeApi: CrowdNodeApi
+) : ViewModel() {
+    val navigationCallback = SingleLiveEvent<NavigationRequest>()
+
+    private val _accountAddress = MutableLiveData(getOrCreateAccountAddress())
+    var accountAddress: LiveData<String> = MediatorLiveData<String>().apply {
+        addSource(_accountAddress) {
+            value = it.toBase58()
+        }
+    }
+
+    val needPassphraseBackUp
+        get() = config.remindBackupSeed
+
+    private val _hasEnoughBalance = MutableLiveData<Boolean>()
+    val hasEnoughBalance: LiveData<Boolean>
+        get() = _hasEnoughBalance
+
+    private val _dashBalance = MutableLiveData<Coin>()
+    val dashBalance: LiveData<Coin>
+        get() = _dashBalance
+
+    var crowdNodeSignUpStatus: LiveData<SignUpStatus> = MediatorLiveData<SignUpStatus>().apply {
+        addSource(crowdNodeApi.signUpStatus.asLiveData(), this::setValue)
+        value = crowdNodeApi.signUpStatus.value
+    }
+
+    val crowdNodeError: Exception?
+        get() = crowdNodeApi.apiError
+
+    val termsAccepted = MutableLiveData(false)
+
+    init {
+        walletDataProvider.observeBalance()
+            .distinctUntilChanged()
+            .onEach {
+                _dashBalance.postValue(it)
+                _hasEnoughBalance.postValue(it >= CrowdNodeConstants.MINIMUM_REQUIRED_DASH)
+            }
+            .launchIn(viewModelScope)
+    }
+    
+    fun backupPassphrase() {
+        navigationCallback.postValue(NavigationRequest.BackupPassphrase)
+    }
+
+    fun restoreWallet() {
+        navigationCallback.postValue(NavigationRequest.RestoreWallet)
+    }
+
+    fun buyDash() {
+        navigationCallback.postValue(NavigationRequest.BuyDash)
+    }
+
+    fun sendReport() {
+        navigationCallback.postValue(NavigationRequest.SendReport)
+    }
+
+    suspend fun signUp() {
+        crowdNodeApi.signUp(_accountAddress.value!!)
+    }
+
+    fun reset() {
+        _accountAddress.value = createNewAccountAddress()
+        crowdNodeApi.reset()
+    }
+
+    suspend fun retry() {
+        reset()
+        signUp()
+    }
+
+    fun changeNotifyWhenDone(toNotify: Boolean, intent: Intent?) {
+        crowdNodeApi.showNotificationOnFinished(toNotify, intent)
+    }
+
+    private fun getOrCreateAccountAddress(): Address {
+        val existingAddress = crowdNodeApi.existingAccountAddress
+
+        if (existingAddress != null) {
+            config.crowdNodeAccountAddress = existingAddress.toBase58()
+            return existingAddress
+        }
+
+        val savedAddress = config.crowdNodeAccountAddress
+
+        return if (savedAddress.isNullOrEmpty()) {
+            return createNewAccountAddress()
+        } else {
+            Address.fromString(Context.get().params, savedAddress)
```

Comment:
We probably should.  `Context.get()` will fail with an exception if called on a thread where `Context.propagate()` has not been called.  Since this is called in a ViewModel, then its probably called in a user thread, which is fine.

if in the future we want to support multiple networks in the same app, then `Context.get().params` may return the wrong network.  So its better that we use the network of the wallet.